{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": [],
      "authorship_tag": "ABX9TyNgUg/CdGBuGZd8sjF+rsg2",
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/cellatlas/cellatlas/blob/main/examples/tag-clicktag/preprocess.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!wget --quiet --show-progress https://github.com/stedolan/jq/releases/download/jq-1.6/jq-linux64\n",
        "!chmod +x jq-linux64 && mv jq-linux64 /usr/local/bin/jq"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "DCtbJUnW6Qta",
        "outputId": "621063fd-4496-40f2-a488-ad838c2afdcc"
      },
      "execution_count": 1,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\rjq-linux64            0%[                    ]       0  --.-KB/s               \rjq-linux64          100%[===================>]   3.77M  --.-KB/s    in 0.03s   \n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 2,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "OEii54Z91IKD",
        "outputId": "18c5e4db-a11e-4774-8ce5-014ef35cb14d"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Cloning into 'cellatlas'...\n",
            "remote: Enumerating objects: 198, done.\u001b[K\n",
            "remote: Counting objects: 100% (73/73), done.\u001b[K\n",
            "remote: Compressing objects: 100% (33/33), done.\u001b[K\n",
            "remote: Total 198 (delta 43), reused 69 (delta 39), pack-reused 125\u001b[K\n",
            "Receiving objects: 100% (198/198), 527.02 MiB | 18.98 MiB/s, done.\n",
            "Resolving deltas: 100% (82/82), done.\n",
            "Updating files: 100% (51/51), done.\n"
          ]
        }
      ],
      "source": [
        "!git clone https://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@github.com/cellatlas/cellatlas.git > /dev/null"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!cd cellatlas && pip install --quiet . > /dev/null\n",
        "!yes | pip uninstall --quiet seqspec\n",
        "!pip install --quiet git+https://github.com/IGVF/seqspec.git > /dev/null\n",
        "!pip install --quiet gget kb-python > /dev/null"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "2VTXQLMD2Y6m",
        "outputId": "61400672-0a92-4dad-fc97-53fc340121a6"
      },
      "execution_count": 3,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Proceed (Y/n)? "
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!mv cellatlas/examples/tag-clicktag/* .\n",
        "!gunzip 737K-august-2016.txt.gz"
      ],
      "metadata": {
        "id": "vbIt2r2A3hbP"
      },
      "execution_count": 4,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Preprocessing"
      ],
      "metadata": {
        "id": "1ievY-64-u43"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "### Example the spec"
      ],
      "metadata": {
        "id": "OrUgMsqp-weB"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!seqspec print spec.yaml"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "oxK70t3e718z",
        "outputId": "381af20d-3e3d-4116-9235-29277afa1995"
      },
      "execution_count": 5,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "                                             ┌─'barcode:16'\n",
            "                              ┌─R1.fastq.gz──┤\n",
            "────────────── ──tag──────────┤              └─'umi:10'\n",
            "                              └─R2.fastq.gz─ ──'tag:90'\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "### Fetch the references"
      ],
      "metadata": {
        "id": "OExrOQ4f-xoy"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!gget ref -o ref.json -w dna,gtf homo_sapiens"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "I63OT0t92weV",
        "outputId": "af25ee0d-b8df-4c63-e1ee-f281aef283ce"
      },
      "execution_count": 6,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Tue May  9 07:02:45 2023 INFO Fetching reference information for homo_sapiens from Ensembl release: 109.\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "FA=!echo $(jq -r '.homo_sapiens.genome_dna.ftp' ref.json)\n",
        "FA=FA[0]\n",
        "GTF=!echo $(jq -r '.homo_sapiens.annotation_gtf.ftp' ref.json)\n",
        "GTF=GTF[0]"
      ],
      "metadata": {
        "id": "xpHlsfEA8x1m"
      },
      "execution_count": 7,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "### Build the pipeline"
      ],
      "metadata": {
        "id": "hAa6kDU5-zak"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!cellatlas build -o out \\\n",
        "-m tag \\\n",
        "-s spec.yaml \\\n",
        "-fa $FA -g $GTF \\\n",
        "-fb feature_barcodes.txt \\\n",
        "fastqs/R1.fastq.gz fastqs/R2.fastq.gz"
      ],
      "metadata": {
        "id": "53NxlmVO2Ob_"
      },
      "execution_count": 8,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "!jq  -r '.commands[] | values[] | join(\"\\n\")' out/cellatlas_info.json "
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "049xtISb3sh4",
        "outputId": "783e8ef1-dbbb-4ac3-b1c5-8a61c2db10b8"
      },
      "execution_count": 9,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "kb ref --workflow kite -i out/index.idx -g out/t2g.txt -f1 out/transcriptome.fa feature_barcodes.txt\n",
            "kb count --workflow kite -i out/index.idx -g out/t2g.txt -x 0,0,16:0,16,26:1,0,90 -w 737K-august-2016.txt -o out --h5ad -t 2 fastqs/R1.fastq.gz fastqs/R2.fastq.gz\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "### Run the pipeline"
      ],
      "metadata": {
        "id": "FNidpYNN-10f"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!jq  -r '.commands[] | values[] | join(\"\\n\")' out/cellatlas_info.json | bash"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "7PDLF2Ps6wfz",
        "outputId": "9d50112f-d76b-45f4-c9a5-98f7f5357d44"
      },
      "execution_count": 10,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "[2023-05-09 07:02:53,811]    INFO [ref_kite] Generating mismatch FASTA at out/transcriptome.fa\n",
            "[2023-05-09 07:02:53,828]    INFO [ref_kite] Creating transcript-to-gene mapping at out/t2g.txt\n",
            "[2023-05-09 07:02:53,833]    INFO [ref_kite] Indexing out/transcriptome.fa to out/index.idx\n",
            "[2023-05-09 07:03:04,042]    INFO [count] Using index out/index.idx to generate BUS file to out from\n",
            "[2023-05-09 07:03:04,042]    INFO [count]         fastqs/R1.fastq.gz\n",
            "[2023-05-09 07:03:04,042]    INFO [count]         fastqs/R2.fastq.gz\n",
            "[2023-05-09 07:03:14,680]    INFO [count] Sorting BUS file out/output.bus to out/tmp/output.s.bus\n",
            "[2023-05-09 07:03:20,905]    INFO [count] Inspecting BUS file out/tmp/output.s.bus\n",
            "[2023-05-09 07:03:22,809]    INFO [count] Correcting BUS records in out/tmp/output.s.bus to out/tmp/output.s.c.bus with whitelist 737K-august-2016.txt\n",
            "[2023-05-09 07:03:26,523]    INFO [count] Sorting BUS file out/tmp/output.s.c.bus to out/output.unfiltered.bus\n",
            "[2023-05-09 07:03:32,141]    INFO [count] Generating count matrix out/counts_unfiltered/cells_x_features from BUS file out/output.unfiltered.bus\n",
            "[2023-05-09 07:03:33,561]    INFO [count] Reading matrix out/counts_unfiltered/cells_x_features.mtx\n",
            "[2023-05-09 07:03:33,860]    INFO [count] Writing matrix to h5ad out/counts_unfiltered/adata.h5ad\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!cat out/run_info.json"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "zCq2zEJD7cYf",
        "outputId": "13794e1b-94fa-4198-e7fe-8dd55f59e9bb"
      },
      "execution_count": 11,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "{\n",
            "\t\"n_targets\": 0,\n",
            "\t\"n_bootstraps\": 0,\n",
            "\t\"n_processed\": 1000000,\n",
            "\t\"n_pseudoaligned\": 966777,\n",
            "\t\"n_unique\": 966777,\n",
            "\t\"p_pseudoaligned\": 96.7,\n",
            "\t\"p_unique\": 96.7,\n",
            "\t\"kallisto_version\": \"0.48.0\",\n",
            "\t\"index_version\": 0,\n",
            "\t\"start_time\": \"Tue May  9 07:03:04 2023\",\n",
            "\t\"call\": \"/usr/local/lib/python3.10/dist-packages/kb_python/bins/linux/kallisto/kallisto bus -i out/index.idx -o out -x 0,0,16:0,16,26:1,0,90 -t 2 fastqs/R1.fastq.gz fastqs/R2.fastq.gz\"\n",
            "}\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!cat out/kb_info.json"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "AS8SyOhmH7LB",
        "outputId": "da5cd435-9fc9-4f75-e7fc-3123f2597a53"
      },
      "execution_count": 12,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "{\n",
            "    \"workdir\": \"/content\",\n",
            "    \"version\": \"0.27.3\",\n",
            "    \"kallisto\": {\n",
            "        \"path\": \"/usr/local/lib/python3.10/dist-packages/kb_python/bins/linux/kallisto/kallisto\",\n",
            "        \"version\": \"0.48.0\"\n",
            "    },\n",
            "    \"bustools\": {\n",
            "        \"path\": \"/usr/local/lib/python3.10/dist-packages/kb_python/bins/linux/bustools/bustools\",\n",
            "        \"version\": \"0.41.0\"\n",
            "    },\n",
            "    \"start_time\": \"2023-05-09T07:03:01.936129\",\n",
            "    \"end_time\": \"2023-05-09T07:03:33.912357\",\n",
            "    \"elapsed\": 31.976228,\n",
            "    \"call\": \"/usr/local/bin/kb count --workflow kite -i out/index.idx -g out/t2g.txt -x 0,0,16:0,16,26:1,0,90 -w 737K-august-2016.txt -o out --h5ad -t 2 fastqs/R1.fastq.gz fastqs/R2.fastq.gz\",\n",
            "    \"commands\": [\n",
            "        \"kallisto bus -i out/index.idx -o out -x 0,0,16:0,16,26:1,0,90 -t 2 fastqs/R1.fastq.gz fastqs/R2.fastq.gz\",\n",
            "        \"bustools inspect out/output.bus\",\n",
            "        \"bustools sort -o out/tmp/output.s.bus -T out/tmp -t 2 -m 4G out/output.bus\",\n",
            "        \"bustools inspect out/tmp/output.s.bus\",\n",
            "        \"bustools inspect -o out/inspect.json -w 737K-august-2016.txt out/tmp/output.s.bus\",\n",
            "        \"bustools correct -o out/tmp/output.s.c.bus -w 737K-august-2016.txt out/tmp/output.s.bus\",\n",
            "        \"bustools inspect out/tmp/output.s.c.bus\",\n",
            "        \"bustools sort -o out/output.unfiltered.bus -T out/tmp -t 2 -m 4G out/tmp/output.s.c.bus\",\n",
            "        \"bustools inspect out/output.unfiltered.bus\",\n",
            "        \"bustools count -o out/counts_unfiltered/cells_x_features -g out/t2g.txt -e out/matrix.ec -t out/transcripts.txt --genecounts out/output.unfiltered.bus\"\n",
            "    ],\n",
            "    \"runtimes\": [\n",
            "        9.030354976654053,\n",
            "        1.6072075366973877,\n",
            "        4.821167707443237,\n",
            "        1.4027824401855469,\n",
            "        1.9038100242614746,\n",
            "        2.2048768997192383,\n",
            "        1.5082874298095703,\n",
            "        4.213881254196167,\n",
            "        1.40315842628479,\n",
            "        1.202362060546875\n",
            "    ]\n",
            "}"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "H5mVaYkTIT4x"
      },
      "execution_count": 12,
      "outputs": []
    }
  ]
}