{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": [],
      "authorship_tag": "ABX9TyMiEgtgBo9dYin9m07SJN5h",
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/cellatlas/cellatlas/blob/main/examples/crispr-10xcrispr/preprocess.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Building Count Matrices with `cellatlas`\n",
        "\n",
        "A major challenge in uniformly preprocessing large amounts of single-cell genomics data from a variety of different assays is identifying and handling sequenced elements in a coherent and consistent fashion. Cell barcodes in reads from RNAseq data from 10x Multiome, for example, must be extracted and error corrected in the manner as cell barcodes in reads from ATACseq data from 10xMultiome so that barcode-barcode registration can occur. Uniform processing in this way minimzes computational variability and enables cross-assay comparisons.\n",
        "\n",
        "In this notebook we demonstrate how single-cell genomics data can be preprocessed to generate a cell by feature count matrix. This requires:\n",
        "\n",
        "1. FASTQ files\n",
        "2. `seqspec` specification for the FASTQ files\n",
        "3. Genome Sequence FASTA\n",
        "4. Genome Annotation GTF\n",
        "5. (optional) Feature barcode list"
      ],
      "metadata": {
        "id": "l2-LZ9798Ibb"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "-------------------"
      ],
      "metadata": {
        "id": "-QBK2zWg_AI1"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Install packages"
      ],
      "metadata": {
        "id": "iKTmVa2A_BNr"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# Install `tree` to view files\n",
        "!wget --quiet --show-progress ftp://mama.indstate.edu/linux/tree/tree-2.1.0.tgz\n",
        "!tar -xf tree-2.1.0.tgz && cd tree-2.1.0 && make -j16 && make install > /dev/null\n",
        "\n",
        "# Install `jq, a command-line tool for extracting key value pairs from JSON files \n",
        "!wget --quiet --show-progress https://github.com/stedolan/jq/releases/download/jq-1.6/jq-linux64\n",
        "!chmod +x jq-linux64 && mv jq-linux64 /usr/local/bin/jq\n",
        "\n",
        "# Clone the cellatlas repo and install the package\n",
        "!git clone https://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@github.com/cellatlas/cellatlas.git > /dev/null\n",
        "!cd cellatlas && pip install --quiet . > /dev/null\n",
        "\n",
        "# Install dependencies\n",
        "!yes | pip uninstall --quiet seqspec\n",
        "!pip install --quiet git+https://github.com/IGVF/seqspec.git > /dev/null\n",
        "!pip install --quiet gget kb-python > /dev/null"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "2VTXQLMD2Y6m",
        "outputId": "53dcb111-0c5d-48c2-8b65-58143c567d80"
      },
      "execution_count": 1,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\rtree-2.1.0.tgz        0%[                    ]       0  --.-KB/s               \rtree-2.1.0.tgz      100%[===================>]  57.80K  --.-KB/s    in 0.1s    \n",
            "gcc -O3 -std=c11 -pedantic -Wall -D_LARGEFILE64_SOURCE -D_FILE_OFFSET_BITS=64 -c -o tree.o tree.c\n",
            "gcc -O3 -std=c11 -pedantic -Wall -D_LARGEFILE64_SOURCE -D_FILE_OFFSET_BITS=64 -c -o list.o list.c\n",
            "gcc -O3 -std=c11 -pedantic -Wall -D_LARGEFILE64_SOURCE -D_FILE_OFFSET_BITS=64 -c -o hash.o hash.c\n",
            "gcc -O3 -std=c11 -pedantic -Wall -D_LARGEFILE64_SOURCE -D_FILE_OFFSET_BITS=64 -c -o color.o color.c\n",
            "gcc -O3 -std=c11 -pedantic -Wall -D_LARGEFILE64_SOURCE -D_FILE_OFFSET_BITS=64 -c -o file.o file.c\n",
            "gcc -O3 -std=c11 -pedantic -Wall -D_LARGEFILE64_SOURCE -D_FILE_OFFSET_BITS=64 -c -o filter.o filter.c\n",
            "gcc -O3 -std=c11 -pedantic -Wall -D_LARGEFILE64_SOURCE -D_FILE_OFFSET_BITS=64 -c -o info.o info.c\n",
            "gcc -O3 -std=c11 -pedantic -Wall -D_LARGEFILE64_SOURCE -D_FILE_OFFSET_BITS=64 -c -o unix.o unix.c\n",
            "gcc -O3 -std=c11 -pedantic -Wall -D_LARGEFILE64_SOURCE -D_FILE_OFFSET_BITS=64 -c -o xml.o xml.c\n",
            "gcc -O3 -std=c11 -pedantic -Wall -D_LARGEFILE64_SOURCE -D_FILE_OFFSET_BITS=64 -c -o json.o json.c\n",
            "gcc -O3 -std=c11 -pedantic -Wall -D_LARGEFILE64_SOURCE -D_FILE_OFFSET_BITS=64 -c -o html.o html.c\n",
            "gcc -O3 -std=c11 -pedantic -Wall -D_LARGEFILE64_SOURCE -D_FILE_OFFSET_BITS=64 -c -o strverscmp.o strverscmp.c\n",
            "gcc  -o tree tree.o list.o hash.o color.o file.o filter.o info.o unix.o xml.o json.o html.o strverscmp.o\n",
            "jq-linux64          100%[===================>]   3.77M  --.-KB/s    in 0.06s   \n",
            "Cloning into 'cellatlas'...\n",
            "remote: Enumerating objects: 315, done.\u001b[K\n",
            "remote: Counting objects: 100% (60/60), done.\u001b[K\n",
            "remote: Compressing objects: 100% (46/46), done.\u001b[K\n",
            "remote: Total 315 (delta 29), reused 30 (delta 14), pack-reused 255\u001b[K\n",
            "Receiving objects: 100% (315/315), 627.63 MiB | 38.48 MiB/s, done.\n",
            "Resolving deltas: 100% (158/158), done.\n",
            "Updating files: 100% (68/68), done.\n",
            "Proceed (Y/n)? "
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "**Note:** We move the relevant data to the working directory and `gunzip` the barcode onlist."
      ],
      "metadata": {
        "id": "6RhzGyP9_cEu"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!mv cellatlas/examples/crispr-10xcrispr/* .\n",
        "!gunzip 3M-february-2018.txt.gz"
      ],
      "metadata": {
        "id": "vbIt2r2A3hbP"
      },
      "execution_count": 2,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Preprocessing"
      ],
      "metadata": {
        "id": "1ievY-64-u43"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "### Example the spec\n",
        "\n",
        "We first use `seqspec print` to check that the read structure matches what we expect. This command prints out an ordered tree representation of the sequenced elements contained in the FASTQ files. Note that the names of the nodes in the `seqspec` must match the names of the FASTQ files."
      ],
      "metadata": {
        "id": "OrUgMsqp-weB"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!seqspec print spec.yaml"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "oxK70t3e718z",
        "outputId": "da97e03e-2051-4058-af12-f77b2a8047f8"
      },
      "execution_count": 3,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "                                             ┌─'barcode:16'\n",
            "                              ┌─R1.fastq.gz──┤\n",
            "────────────── ──crispr───────┤              └─'umi:12'\n",
            "                              └─R2.fastq.gz─ ──'gDNA:90'\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "### Fetch the references\n",
        "This step is only necessary if modality that we are processing uses a transcriptome reference-based alignment."
      ],
      "metadata": {
        "id": "OExrOQ4f-xoy"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!gget ref -o ref.json -w dna,gtf homo_sapiens"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "I63OT0t92weV",
        "outputId": "45b32979-5478-4cb8-e0bb-3561e631c18d"
      },
      "execution_count": 4,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Thu Jun  1 09:10:18 2023 INFO Fetching reference information for homo_sapiens from Ensembl release: 109.\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "FA=!echo $(jq -r '.homo_sapiens.genome_dna.ftp' ref.json)\n",
        "FA=FA[0]\n",
        "GTF=!echo $(jq -r '.homo_sapiens.annotation_gtf.ftp' ref.json)\n",
        "GTF=GTF[0]"
      ],
      "metadata": {
        "id": "xpHlsfEA8x1m"
      },
      "execution_count": 5,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "### Build the pipeline\n",
        "\n",
        "We now supply all of the relevant objects to `cellatlas build` to produce the appropriate commands to be run to build the pipeline. This includes a reference building step and a read counting and quantification step both of which are performed with `kallisto` and `bustools` as part of the `kb-python` package."
      ],
      "metadata": {
        "id": "hAa6kDU5-zak"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!cellatlas build -o out \\\n",
        "-m crispr \\\n",
        "-s spec.yaml \\\n",
        "-fa $FA -g $GTF \\\n",
        "-fb feature_barcodes.txt \\\n",
        "fastqs/R1.fastq.gz fastqs/R2.fastq.gz"
      ],
      "metadata": {
        "id": "53NxlmVO2Ob_"
      },
      "execution_count": 10,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "**Note**: The commands generated by `cellatlas build` are stored in the `out/cellatlas_info.json` file. We can view this file with `jq`."
      ],
      "metadata": {
        "id": "2CvtZOUDAjfN"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!jq  -r '.commands[] | values[] | join(\"\\n\")' out/cellatlas_info.json "
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "049xtISb3sh4",
        "outputId": "961dc55e-c0d9-4da5-ae04-6f1b972f275d"
      },
      "execution_count": 11,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "kb ref --workflow kite -i out/index.idx -g out/t2g.txt -f1 out/transcriptome.fa feature_barcodes.txt\n",
            "kb count --workflow kite -i out/index.idx -g out/t2g.txt -x 0,0,16:0,16,28:1,0,90 -w 3M-february-2018.txt -o out --h5ad -t 2 fastqs/R1.fastq.gz fastqs/R2.fastq.gz\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "### Run the pipeline\n",
        "\n",
        "To run the pipeline we simply extract the commands from `out/cellatlas_info.json` and pass them to `bash`. "
      ],
      "metadata": {
        "id": "FNidpYNN-10f"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!jq  -r '.commands[] | values[] | join(\"\\n\")' out/cellatlas_info.json | bash"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "7PDLF2Ps6wfz",
        "outputId": "7fe0c472-4839-4cd2-85ba-130bbfb0ae47"
      },
      "execution_count": 17,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "[2023-06-01 09:23:02,865]    INFO [ref_kite] Generating mismatch FASTA at out/transcriptome.fa\n",
            "[2023-06-01 09:23:02,883] WARNING [ref_kite] Detected barcodes of different lengths: 19,20\n",
            "[2023-06-01 09:23:02,924]    INFO [ref_kite] Creating transcript-to-gene mapping at out/t2g.txt\n",
            "[2023-06-01 09:23:02,951]    INFO [ref_kite] Indexing out/transcriptome.fa to out/index.idx\n",
            "[2023-06-01 09:23:12,759]    INFO [count] Using index out/index.idx to generate BUS file to out from\n",
            "[2023-06-01 09:23:12,759]    INFO [count]         fastqs/R1.fastq.gz\n",
            "[2023-06-01 09:23:12,759]    INFO [count]         fastqs/R2.fastq.gz\n",
            "[2023-06-01 09:23:22,891]    INFO [count] Sorting BUS file out/output.bus to out/tmp/output.s.bus\n",
            "[2023-06-01 09:23:29,505]    INFO [count] Inspecting BUS file out/tmp/output.s.bus\n",
            "[2023-06-01 09:23:35,516]    INFO [count] Correcting BUS records in out/tmp/output.s.bus to out/tmp/output.s.c.bus with whitelist 3M-february-2018.txt\n",
            "[2023-06-01 09:23:48,351]    INFO [count] Sorting BUS file out/tmp/output.s.c.bus to out/output.unfiltered.bus\n",
            "[2023-06-01 09:23:53,663]    INFO [count] Generating count matrix out/counts_unfiltered/cells_x_features from BUS file out/output.unfiltered.bus\n",
            "[2023-06-01 09:23:55,115]    INFO [count] Reading matrix out/counts_unfiltered/cells_x_features.mtx\n",
            "[2023-06-01 09:23:55,446]    INFO [count] Writing matrix to h5ad out/counts_unfiltered/adata.h5ad\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "### Inspect outputs\n",
        "\n",
        "We inspect the `out/run_info.json` and `out/kb_info.json` as a simple QC on the pipeline."
      ],
      "metadata": {
        "id": "z5PcMU52A1DE"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!cat out/run_info.json"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "zCq2zEJD7cYf",
        "outputId": "91349dbc-8ea4-4381-d056-febecc87d65e"
      },
      "execution_count": 18,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "{\n",
            "\t\"n_targets\": 0,\n",
            "\t\"n_bootstraps\": 0,\n",
            "\t\"n_processed\": 1000000,\n",
            "\t\"n_pseudoaligned\": 779588,\n",
            "\t\"n_unique\": 779588,\n",
            "\t\"p_pseudoaligned\": 78.0,\n",
            "\t\"p_unique\": 78.0,\n",
            "\t\"kallisto_version\": \"0.48.0\",\n",
            "\t\"index_version\": 0,\n",
            "\t\"start_time\": \"Thu Jun  1 09:23:12 2023\",\n",
            "\t\"call\": \"/usr/local/lib/python3.10/dist-packages/kb_python/bins/linux/kallisto/kallisto bus -i out/index.idx -o out -x 0,0,16:0,16,28:1,0,90 -t 2 fastqs/R1.fastq.gz fastqs/R2.fastq.gz\"\n",
            "}\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!cat out/kb_info.json"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "AS8SyOhmH7LB",
        "outputId": "85a3456d-7bc1-47af-a90a-d97060967454"
      },
      "execution_count": 19,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "{\n",
            "    \"workdir\": \"/content\",\n",
            "    \"version\": \"0.27.3\",\n",
            "    \"kallisto\": {\n",
            "        \"path\": \"/usr/local/lib/python3.10/dist-packages/kb_python/bins/linux/kallisto/kallisto\",\n",
            "        \"version\": \"0.48.0\"\n",
            "    },\n",
            "    \"bustools\": {\n",
            "        \"path\": \"/usr/local/lib/python3.10/dist-packages/kb_python/bins/linux/bustools/bustools\",\n",
            "        \"version\": \"0.41.0\"\n",
            "    },\n",
            "    \"start_time\": \"2023-06-01T09:23:10.653352\",\n",
            "    \"end_time\": \"2023-06-01T09:23:55.515173\",\n",
            "    \"elapsed\": 44.861821,\n",
            "    \"call\": \"/usr/local/bin/kb count --workflow kite -i out/index.idx -g out/t2g.txt -x 0,0,16:0,16,28:1,0,90 -w 3M-february-2018.txt -o out --h5ad -t 2 fastqs/R1.fastq.gz fastqs/R2.fastq.gz\",\n",
            "    \"commands\": [\n",
            "        \"kallisto bus -i out/index.idx -o out -x 0,0,16:0,16,28:1,0,90 -t 2 fastqs/R1.fastq.gz fastqs/R2.fastq.gz\",\n",
            "        \"bustools inspect out/output.bus\",\n",
            "        \"bustools sort -o out/tmp/output.s.bus -T out/tmp -t 2 -m 4G out/output.bus\",\n",
            "        \"bustools inspect out/tmp/output.s.bus\",\n",
            "        \"bustools inspect -o out/inspect.json -w 3M-february-2018.txt out/tmp/output.s.bus\",\n",
            "        \"bustools correct -o out/tmp/output.s.c.bus -w 3M-february-2018.txt out/tmp/output.s.bus\",\n",
            "        \"bustools inspect out/tmp/output.s.c.bus\",\n",
            "        \"bustools sort -o out/output.unfiltered.bus -T out/tmp -t 2 -m 4G out/tmp/output.s.c.bus\",\n",
            "        \"bustools inspect out/output.unfiltered.bus\",\n",
            "        \"bustools count -o out/counts_unfiltered/cells_x_features -g out/t2g.txt -e out/matrix.ec -t out/transcripts.txt --genecounts out/output.unfiltered.bus\"\n",
            "    ],\n",
            "    \"runtimes\": [\n",
            "        8.728450298309326,\n",
            "        1.4028401374816895,\n",
            "        5.211154460906982,\n",
            "        1.4026970863342285,\n",
            "        6.010495185852051,\n",
            "        11.431258201599121,\n",
            "        1.4028937816619873,\n",
            "        3.9081506729125977,\n",
            "        1.4028990268707275,\n",
            "        1.201439380645752\n",
            "    ]\n",
            "}"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!tree out"
      ],
      "metadata": {
        "id": "H5mVaYkTIT4x",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "2884ca6d-b1f9-4081-edfc-00372b0d32cd"
      },
      "execution_count": 20,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\u001b[01;34mout\u001b[0m\n",
            "├── \u001b[00mcellatlas_info.json\u001b[0m\n",
            "├── \u001b[01;34mcounts_unfiltered\u001b[0m\n",
            "│   ├── \u001b[00madata.h5ad\u001b[0m\n",
            "│   ├── \u001b[00mcells_x_features.barcodes.txt\u001b[0m\n",
            "│   ├── \u001b[00mcells_x_features.genes.txt\u001b[0m\n",
            "│   └── \u001b[00mcells_x_features.mtx\u001b[0m\n",
            "├── \u001b[00mindex.idx\u001b[0m\n",
            "├── \u001b[00minspect.json\u001b[0m\n",
            "├── \u001b[00mkb_info.json\u001b[0m\n",
            "├── \u001b[00mmatrix.ec\u001b[0m\n",
            "├── \u001b[00moutput.bus\u001b[0m\n",
            "├── \u001b[00moutput.unfiltered.bus\u001b[0m\n",
            "├── \u001b[00mrun_info.json\u001b[0m\n",
            "├── \u001b[00mt2g.txt\u001b[0m\n",
            "├── \u001b[00mtranscriptome.fa\u001b[0m\n",
            "└── \u001b[00mtranscripts.txt\u001b[0m\n",
            "\n",
            "2 directories, 15 files\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "-udjiJdTRWu6"
      },
      "execution_count": null,
      "outputs": []
    }
  ]
}