{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": [],
      "authorship_tag": "ABX9TyMN9+MPVwTzyIuV9+3c/EWu",
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/cellatlas/cellatlas/blob/main/examples/atac-10xatac/preprocess.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Building Count Matrices with `cellatlas`\n",
        "\n",
        "A major challenge in uniformly preprocessing large amounts of single-cell genomics data from a variety of different assays is identifying and handling sequenced elements in a coherent and consistent fashion. Cell barcodes in reads from RNAseq data from 10x Multiome, for example, must be extracted and error corrected in the manner as cell barcodes in reads from ATACseq data from 10xMultiome so that barcode-barcode registration can occur. Uniform processing in this way minimzes computational variability and enables cross-assay comparisons.\n",
        "\n",
        "In this notebook we demonstrate how single-cell genomics data can be preprocessed to generate a cell by feature count matrix. This requires:\n",
        "\n",
        "1. FASTQ files\n",
        "2. `seqspec` specification for the FASTQ files\n",
        "3. Genome Sequence FASTA\n",
        "4. Genome Annotation GTF\n",
        "5. (optional) Feature barcode list"
      ],
      "metadata": {
        "id": "l2-LZ9798Ibb"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "-------------------"
      ],
      "metadata": {
        "id": "-QBK2zWg_AI1"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Install packages"
      ],
      "metadata": {
        "id": "iKTmVa2A_BNr"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# Install `tree` to view files\n",
        "!wget --quiet --show-progress ftp://mama.indstate.edu/linux/tree/tree-2.1.0.tgz\n",
        "!tar -xf tree-2.1.0.tgz && cd tree-2.1.0 && make -j16 && make install > /dev/null\n",
        "\n",
        "# Install `jq, a command-line tool for extracting key value pairs from JSON files \n",
        "!wget --quiet --show-progress https://github.com/stedolan/jq/releases/download/jq-1.6/jq-linux64\n",
        "!chmod +x jq-linux64 && mv jq-linux64 /usr/local/bin/jq\n",
        "\n",
        "# Clone the cellatlas repo and install the package\n",
        "!git clone https://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@github.com/cellatlas/cellatlas.git > /dev/null\n",
        "!cd cellatlas && pip install --quiet . > /dev/null\n",
        "\n",
        "# Install dependencies\n",
        "!yes | pip uninstall --quiet seqspec\n",
        "!pip install --quiet git+https://github.com/IGVF/seqspec.git > /dev/null\n",
        "!pip install --quiet gget kb-python > /dev/null"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "2VTXQLMD2Y6m",
        "outputId": "f6d2b729-0c4e-4cd1-9f50-30d483a2874e"
      },
      "execution_count": 1,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\rtree-2.1.0.tgz        0%[                    ]       0  --.-KB/s               \rtree-2.1.0.tgz      100%[===================>]  57.80K  --.-KB/s    in 0.05s   \n",
            "gcc -O3 -std=c11 -pedantic -Wall -D_LARGEFILE64_SOURCE -D_FILE_OFFSET_BITS=64 -c -o tree.o tree.c\n",
            "gcc -O3 -std=c11 -pedantic -Wall -D_LARGEFILE64_SOURCE -D_FILE_OFFSET_BITS=64 -c -o list.o list.c\n",
            "gcc -O3 -std=c11 -pedantic -Wall -D_LARGEFILE64_SOURCE -D_FILE_OFFSET_BITS=64 -c -o hash.o hash.c\n",
            "gcc -O3 -std=c11 -pedantic -Wall -D_LARGEFILE64_SOURCE -D_FILE_OFFSET_BITS=64 -c -o color.o color.c\n",
            "gcc -O3 -std=c11 -pedantic -Wall -D_LARGEFILE64_SOURCE -D_FILE_OFFSET_BITS=64 -c -o file.o file.c\n",
            "gcc -O3 -std=c11 -pedantic -Wall -D_LARGEFILE64_SOURCE -D_FILE_OFFSET_BITS=64 -c -o filter.o filter.c\n",
            "gcc -O3 -std=c11 -pedantic -Wall -D_LARGEFILE64_SOURCE -D_FILE_OFFSET_BITS=64 -c -o info.o info.c\n",
            "gcc -O3 -std=c11 -pedantic -Wall -D_LARGEFILE64_SOURCE -D_FILE_OFFSET_BITS=64 -c -o unix.o unix.c\n",
            "gcc -O3 -std=c11 -pedantic -Wall -D_LARGEFILE64_SOURCE -D_FILE_OFFSET_BITS=64 -c -o xml.o xml.c\n",
            "gcc -O3 -std=c11 -pedantic -Wall -D_LARGEFILE64_SOURCE -D_FILE_OFFSET_BITS=64 -c -o json.o json.c\n",
            "gcc -O3 -std=c11 -pedantic -Wall -D_LARGEFILE64_SOURCE -D_FILE_OFFSET_BITS=64 -c -o html.o html.c\n",
            "gcc -O3 -std=c11 -pedantic -Wall -D_LARGEFILE64_SOURCE -D_FILE_OFFSET_BITS=64 -c -o strverscmp.o strverscmp.c\n",
            "gcc  -o tree tree.o list.o hash.o color.o file.o filter.o info.o unix.o xml.o json.o html.o strverscmp.o\n",
            "jq-linux64          100%[===================>]   3.77M  --.-KB/s    in 0.02s   \n",
            "Cloning into 'cellatlas'...\n",
            "remote: Enumerating objects: 213, done.\u001b[K\n",
            "remote: Counting objects: 100% (88/88), done.\u001b[K\n",
            "remote: Compressing objects: 100% (48/48), done.\u001b[K\n",
            "remote: Total 213 (delta 55), reused 66 (delta 39), pack-reused 125\u001b[K\n",
            "Receiving objects: 100% (213/213), 527.02 MiB | 40.73 MiB/s, done.\n",
            "Resolving deltas: 100% (94/94), done.\n",
            "Updating files: 100% (52/52), done.\n",
            "Proceed (Y/n)? "
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "**Note:** We move the relevant data to the working directory and `gunzip` the barcode onlist."
      ],
      "metadata": {
        "id": "6RhzGyP9_cEu"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!mv cellatlas/examples/tag-clicktag/* .\n",
        "!gunzip 737K-august-2016.txt.gz"
      ],
      "metadata": {
        "id": "vbIt2r2A3hbP"
      },
      "execution_count": 2,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Preprocessing"
      ],
      "metadata": {
        "id": "1ievY-64-u43"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "### Example the spec\n",
        "\n",
        "We first use `seqspec print` to check that the read structure matches what we expect. This command prints out an ordered tree representation of the sequenced elements contained in the FASTQ files. Note that the names of the nodes in the `seqspec` must match the names of the FASTQ files."
      ],
      "metadata": {
        "id": "OrUgMsqp-weB"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!seqspec print spec.yaml"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "oxK70t3e718z",
        "outputId": "a992a89c-8daf-4d4e-8492-4dad55b0162b"
      },
      "execution_count": 3,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "                                             ┌─'barcode:16'\n",
            "                              ┌─R1.fastq.gz──┤\n",
            "────────────── ──tag──────────┤              └─'umi:10'\n",
            "                              └─R2.fastq.gz─ ──'tag:90'\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "### Fetch the references\n",
        "This step is only necessary if modality that we are processing uses a transcriptome reference-based alignment."
      ],
      "metadata": {
        "id": "OExrOQ4f-xoy"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!gget ref -d -o ref.json -w dna,gtf homo_sapiens"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "I63OT0t92weV",
        "outputId": "2a5f8aa8-5a15-496e-d254-e6056f1171ad"
      },
      "execution_count": 11,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Tue May 16 16:46:02 2023 INFO Fetching reference information for homo_sapiens from Ensembl release: 109.\n",
            "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
            "                                 Dload  Upload   Total   Spent    Left  Speed\n",
            "100  840M  100  840M    0     0   454k      0  0:31:33  0:31:33 --:--:--  464k\n",
            "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
            "                                 Dload  Upload   Total   Spent    Left  Speed\n",
            "100 51.7M  100 51.7M    0     0   449k      0  0:01:58  0:01:58 --:--:--  450k\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "FA=!echo $(jq -r '.homo_sapiens.genome_dna.ftp' ref.json)\n",
        "FA=FA[0]\n",
        "GTF=!echo $(jq -r '.homo_sapiens.annotation_gtf.ftp' ref.json)\n",
        "GTF=GTF[0]"
      ],
      "metadata": {
        "id": "xpHlsfEA8x1m"
      },
      "execution_count": 5,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "### Build the pipeline\n",
        "\n",
        "We now supply all of the relevant objects to `cellatlas build` to produce the appropriate commands to be run to build the pipeline. This includes a reference building step and a read counting and quantification step both of which are performed with `kallisto` and `bustools` as part of the `kb-python` package."
      ],
      "metadata": {
        "id": "hAa6kDU5-zak"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!cellatlas build -o out \\\n",
        "-m atac \\\n",
        "-s spec.yaml \\\n",
        "-fa Homo_sapiens.GRCh38.dna.primary_assembly.fa.gz -g Homo_sapiens.GRCh38.109.gtf.gz \\\n",
        "-fb feature_barcodes.txt \\\n",
        "fastqs/I2.fastq.gz fastqs/R1.fastq.gz fastqs/R2.fastq.gz"
      ],
      "metadata": {
        "id": "53NxlmVO2Ob_"
      },
      "execution_count": 13,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "**Note**: The commands generated by `cellatlas build` are stored in the `out/cellatlas_info.json` file. We can view this file with `jq`."
      ],
      "metadata": {
        "id": "2CvtZOUDAjfN"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!jq  -r '.commands[] | values[] | join(\"\\n\")' out/cellatlas_info.json "
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "049xtISb3sh4",
        "outputId": "3790ca86-8422-44cd-863d-e831f5f13a4f"
      },
      "execution_count": 14,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "minimap2 -d out/ref.mmi Homo_sapiens.GRCh38.dna.primary_assembly.fa.gz\n",
            "gunzip Homo_sapiens.GRCh38.dna.primary_assembly.fa.gz | fold -w 80 > out/genome.fa\n",
            "minimap2 -o out/genome.sam -a -x sr -t 32 out/ref.mmi  \n",
            "samtools view -@ 8 -o out/genome.u.bam -b genome.sam\n",
            "samtools sort -@ 8 -o out/genome.bam -n -m 8G out/genome.u.bam\n",
            "Genrich -t out/genome.bam -o out/genome.bed -f out/genome_peaks.log -v\n",
            "cat out/genome.bed | bedtools sort | bedtools merge > out/peaks.bed\n",
            "bedtools getfasta -fi out/genome.fa -bed out/peaks.bed -fo out/peaks.fa\n",
            "cat out/peaks.fa | awk '{if($1~/>/)print $1\"\t\"$1\"\t\"$1}' > out/t2g.txt\n",
            "sed -i 's/>//g' out/t2g.txt\n",
            "kallisto index -i out/peaks.idx out/peaks.fa\n",
            "kb count --workflow kite -i out/index.idx -g out/t2g.txt -x 0,0,16:-1,-1,-1:1,0,50,2,0,50 -w ATA-737K-cratac-v1.txt -o out --h5ad -t 2 fastqs/I2.fastq.gz fastqs/R1.fastq.gz fastqs/R2.fastq.gz\n",
            "mkdir -p out/counts_mult\n",
            "bustools count -o out/counts_mult/cells_x_genes -g out/t2g.txt -e out/matrix.ec -t out/transcripts.txt --genecounts --cm out/output.unfiltered.bus\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "### Run the pipeline\n",
        "\n",
        "To run the pipeline we simply extract the commands from `out/cellatlas_info.json` and pass them to `bash`. "
      ],
      "metadata": {
        "id": "FNidpYNN-10f"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!jq  -r '.commands[] | values[] | join(\"\\n\")' out/cellatlas_info.json | bash"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "7PDLF2Ps6wfz",
        "outputId": "c102f2c8-8d5d-4b40-ff3a-dd59c0e5bf88"
      },
      "execution_count": 15,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "bash: line 1: 39408 Killed                  minimap2 -d out/ref.mmi Homo_sapiens.GRCh38.dna.primary_assembly.fa.gz\n",
            "[ERROR] missing input: please specify a query file to map or option -d to keep the index\n",
            "[E::hts_open_format] Failed to open file \"genome.sam\" : No such file or directory\n",
            "samtools view: failed to open \"genome.sam\" for reading: No such file or directory\n",
            "[E::hts_open_format] Failed to open file \"out/genome.u.bam\" : No such file or directory\n",
            "samtools sort: can't open \"out/genome.u.bam\": No such file or directory\n",
            "Error! out/genome.bam: cannot open file for reading\n",
            "cat: out/genome.bed: No such file or directory\n",
            "index file out/genome.fa.fai not found, generating...\n",
            "terminate called after throwing an instance of 'std::out_of_range'\n",
            "  what():  vector::_M_range_check: __n (which is 0) >= this->size() (which is 0)\n",
            "bash: line 8: 40119 Aborted                 (core dumped) bedtools getfasta -fi out/genome.fa -bed out/peaks.bed -fo out/peaks.fa\n",
            "\n",
            "[build] loading fasta file out/peaks.fa\n",
            "[build] k-mer length: 31\n",
            "[build] counting k-mers ... done.\n",
            "[build] building target de Bruijn graph ...  done \n",
            "[build] creating equivalence classes ...  done\n",
            "[build] target de Bruijn graph has 0 contigs and contains 0 k-mers \n",
            "\n",
            "[2023-05-16 17:27:47,584]    INFO [count] Using index out/index.idx to generate BUS file to out from\n",
            "[2023-05-16 17:27:47,584]    INFO [count]         fastqs/I2.fastq.gz\n",
            "[2023-05-16 17:27:47,585]    INFO [count]         fastqs/R1.fastq.gz\n",
            "[2023-05-16 17:27:47,585]    INFO [count]         fastqs/R2.fastq.gz\n",
            "[2023-05-16 17:27:48,687]   ERROR [count] kallisto 0.48.0\n",
            "Generates BUS files for single-cell sequencing\n",
            "\n",
            "Usage: kallisto bus [arguments] FASTQ-files\n",
            "\n",
            "Required arguments:\n",
            "-i, --index=STRING            Filename for the kallisto index to be used for\n",
            "pseudoalignment\n",
            "-o, --output-dir=STRING       Directory to write output to\n",
            "\n",
            "Optional arguments:\n",
            "-x, --technology=STRING       Single-cell technology used\n",
            "-l, --list                    List all single-cell technologies supported\n",
            "-B, --batch=FILE              Process files listed in FILE\n",
            "-t, --threads=INT             Number of threads to use (default: 1)\n",
            "-b, --bam                     Input file is a BAM file\n",
            "-n, --num                     Output number of read in flag column (incompatible with --bam)\n",
            "-T, --tag=STRING              5′ tag sequence to identify UMI reads for certain technologies\n",
            "--fr-stranded             Strand specific reads for UMI-tagged reads, first read forward\n",
            "--rf-stranded             Strand specific reads for UMI-tagged reads, first read reverse\n",
            "--unstranded              Treat all read as non-strand-specific\n",
            "--paired                  Treat reads as paired\n",
            "--genomebam               Project pseudoalignments to genome sorted BAM file\n",
            "-g, --gtf                     GTF file for transcriptome information\n",
            "(required for --genomebam)\n",
            "-c, --chromosomes             Tab separated file with chromosome names and lengths\n",
            "(optional for --genomebam, but recommended)\n",
            "--verbose                 Print out progress information every 1M proccessed reads\n",
            "\n",
            "Error: kallisto index file not found out/index.idx\n",
            "[2023-05-16 17:27:48,687]   ERROR [main] An exception occurred\n",
            "Traceback (most recent call last):\n",
            "  File \"/usr/local/lib/python3.10/dist-packages/kb_python/main.py\", line 1305, in main\n",
            "    COMMAND_TO_FUNCTION[args.command](parser, args, temp_dir=temp_dir)\n",
            "  File \"/usr/local/lib/python3.10/dist-packages/kb_python/main.py\", line 550, in parse_count\n",
            "    count(\n",
            "  File \"/usr/local/lib/python3.10/dist-packages/ngs_tools/logging.py\", line 62, in inner\n",
            "    return func(*args, **kwargs)\n",
            "  File \"/usr/local/lib/python3.10/dist-packages/kb_python/count.py\", line 1038, in count\n",
            "    bus_result = kallisto_bus(\n",
            "  File \"/usr/local/lib/python3.10/dist-packages/kb_python/validate.py\", line 116, in inner\n",
            "    results = func(*args, **kwargs)\n",
            "  File \"/usr/local/lib/python3.10/dist-packages/kb_python/count.py\", line 150, in kallisto_bus\n",
            "    run_executable(command)\n",
            "  File \"/usr/local/lib/python3.10/dist-packages/kb_python/dry/__init__.py\", line 25, in inner\n",
            "    return func(*args, **kwargs)\n",
            "  File \"/usr/local/lib/python3.10/dist-packages/kb_python/utils.py\", line 203, in run_executable\n",
            "    raise sp.CalledProcessError(p.returncode, ' '.join(command))\n",
            "subprocess.CalledProcessError: Command '/usr/local/lib/python3.10/dist-packages/kb_python/bins/linux/kallisto/kallisto bus -i out/index.idx -o out -x 0,0,16:-1,-1,-1:1,0,50,2,0,50 -t 2 fastqs/I2.fastq.gz fastqs/R1.fastq.gz fastqs/R2.fastq.gz' returned non-zero exit status 1.\n",
            "bash: line 14: bustools: command not found\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "### Inspect outputs\n",
        "\n",
        "We inspect the `out/run_info.json` and `out/kb_info.json` as a simple QC on the pipeline."
      ],
      "metadata": {
        "id": "z5PcMU52A1DE"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!cat out/run_info.json"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "zCq2zEJD7cYf",
        "outputId": "04837748-5f78-4cb9-896a-b1d4e08b0fda"
      },
      "execution_count": 9,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "{\n",
            "\t\"n_targets\": 0,\n",
            "\t\"n_bootstraps\": 0,\n",
            "\t\"n_processed\": 1000000,\n",
            "\t\"n_pseudoaligned\": 966777,\n",
            "\t\"n_unique\": 966777,\n",
            "\t\"p_pseudoaligned\": 96.7,\n",
            "\t\"p_unique\": 96.7,\n",
            "\t\"kallisto_version\": \"0.48.0\",\n",
            "\t\"index_version\": 0,\n",
            "\t\"start_time\": \"Fri May 12 01:57:18 2023\",\n",
            "\t\"call\": \"/usr/local/lib/python3.10/dist-packages/kb_python/bins/linux/kallisto/kallisto bus -i out/index.idx -o out -x 0,0,16:0,16,26:1,0,90 -t 2 fastqs/R1.fastq.gz fastqs/R2.fastq.gz\"\n",
            "}\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!cat out/kb_info.json"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "AS8SyOhmH7LB",
        "outputId": "5295e84d-1f61-49d6-d5df-da20f279dc5d"
      },
      "execution_count": 10,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "{\n",
            "    \"workdir\": \"/content\",\n",
            "    \"version\": \"0.27.3\",\n",
            "    \"kallisto\": {\n",
            "        \"path\": \"/usr/local/lib/python3.10/dist-packages/kb_python/bins/linux/kallisto/kallisto\",\n",
            "        \"version\": \"0.48.0\"\n",
            "    },\n",
            "    \"bustools\": {\n",
            "        \"path\": \"/usr/local/lib/python3.10/dist-packages/kb_python/bins/linux/bustools/bustools\",\n",
            "        \"version\": \"0.41.0\"\n",
            "    },\n",
            "    \"start_time\": \"2023-05-12T01:57:16.346153\",\n",
            "    \"end_time\": \"2023-05-12T01:57:54.493175\",\n",
            "    \"elapsed\": 38.147022,\n",
            "    \"call\": \"/usr/local/bin/kb count --workflow kite -i out/index.idx -g out/t2g.txt -x 0,0,16:0,16,26:1,0,90 -w 737K-august-2016.txt -o out --h5ad -t 2 fastqs/R1.fastq.gz fastqs/R2.fastq.gz\",\n",
            "    \"commands\": [\n",
            "        \"kallisto bus -i out/index.idx -o out -x 0,0,16:0,16,26:1,0,90 -t 2 fastqs/R1.fastq.gz fastqs/R2.fastq.gz\",\n",
            "        \"bustools inspect out/output.bus\",\n",
            "        \"bustools sort -o out/tmp/output.s.bus -T out/tmp -t 2 -m 4G out/output.bus\",\n",
            "        \"bustools inspect out/tmp/output.s.bus\",\n",
            "        \"bustools inspect -o out/inspect.json -w 737K-august-2016.txt out/tmp/output.s.bus\",\n",
            "        \"bustools correct -o out/tmp/output.s.c.bus -w 737K-august-2016.txt out/tmp/output.s.bus\",\n",
            "        \"bustools inspect out/tmp/output.s.c.bus\",\n",
            "        \"bustools sort -o out/output.unfiltered.bus -T out/tmp -t 2 -m 4G out/tmp/output.s.c.bus\",\n",
            "        \"bustools inspect out/output.unfiltered.bus\",\n",
            "        \"bustools count -o out/counts_unfiltered/cells_x_features -g out/t2g.txt -e out/matrix.ec -t out/transcripts.txt --genecounts out/output.unfiltered.bus\"\n",
            "    ],\n",
            "    \"runtimes\": [\n",
            "        10.37208366394043,\n",
            "        1.4030237197875977,\n",
            "        9.364439964294434,\n",
            "        1.404646635055542,\n",
            "        1.7034249305725098,\n",
            "        2.004028081893921,\n",
            "        1.302722692489624,\n",
            "        5.110419034957886,\n",
            "        1.302722692489624,\n",
            "        1.202413558959961\n",
            "    ]\n",
            "}"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!tree out"
      ],
      "metadata": {
        "id": "H5mVaYkTIT4x",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "25ed138a-3dd8-4bc2-bc23-4315fec2cd8a"
      },
      "execution_count": 11,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\u001b[01;34mout\u001b[0m\n",
            "├── \u001b[00mcellatlas_info.json\u001b[0m\n",
            "├── \u001b[01;34mcounts_unfiltered\u001b[0m\n",
            "│   ├── \u001b[00madata.h5ad\u001b[0m\n",
            "│   ├── \u001b[00mcells_x_features.barcodes.txt\u001b[0m\n",
            "│   ├── \u001b[00mcells_x_features.genes.txt\u001b[0m\n",
            "│   └── \u001b[00mcells_x_features.mtx\u001b[0m\n",
            "├── \u001b[00mindex.idx\u001b[0m\n",
            "├── \u001b[00minspect.json\u001b[0m\n",
            "├── \u001b[00mkb_info.json\u001b[0m\n",
            "├── \u001b[00mmatrix.ec\u001b[0m\n",
            "├── \u001b[00moutput.bus\u001b[0m\n",
            "├── \u001b[00moutput.unfiltered.bus\u001b[0m\n",
            "├── \u001b[00mrun_info.json\u001b[0m\n",
            "├── \u001b[00mt2g.txt\u001b[0m\n",
            "├── \u001b[00mtranscriptome.fa\u001b[0m\n",
            "└── \u001b[00mtranscripts.txt\u001b[0m\n",
            "\n",
            "2 directories, 15 files\n"
          ]
        }
      ]
    }
  ]
}