{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "colab_type": "text",
    "id": "view-in-github"
   },
   "source": [
    "<a href=\"https://colab.research.google.com/github/cellatlas/cellatlas/blob/main/examples/rna-splitseq/preprocess.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "l2-LZ9798Ibb"
   },
   "source": [
    "# Building Count Matrices with `cellatlas`\n",
    "\n",
    "A major challenge in uniformly preprocessing large amounts of single-cell genomics data from a variety of different assays is identifying and handling sequenced elements in a coherent and consistent fashion. Cell barcodes in reads from RNAseq data from 10x Multiome, for example, must be extracted and error corrected in the manner as cell barcodes in reads from ATACseq data from 10xMultiome so that barcode-barcode registration can occur. Uniform processing in this way minimzes computational variability and enables cross-assay comparisons.\n",
    "\n",
    "In this notebook we demonstrate how single-cell genomics data can be preprocessed to generate a cell by feature count matrix. This requires:\n",
    "\n",
    "1. FASTQ files\n",
    "2. `seqspec` specification for the FASTQ files\n",
    "3. Genome Sequence FASTA\n",
    "4. Genome Annotation GTF\n",
    "5. (optional) Feature barcode list"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "-QBK2zWg_AI1"
   },
   "source": [
    "-------------------"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "iKTmVa2A_BNr"
   },
   "source": [
    "## Install packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "2VTXQLMD2Y6m",
    "outputId": "e94cce37-6ec1-45f1-8328-0b8673f88630"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "tree-2.1.0.tgz      100%[===================>]  57.80K  --.-KB/s    in 0.04s   \n",
      "gcc -O3 -std=c11 -pedantic -Wall -D_LARGEFILE64_SOURCE -D_FILE_OFFSET_BITS=64 -c -o tree.o tree.c\n",
      "gcc -O3 -std=c11 -pedantic -Wall -D_LARGEFILE64_SOURCE -D_FILE_OFFSET_BITS=64 -c -o list.o list.c\n",
      "gcc -O3 -std=c11 -pedantic -Wall -D_LARGEFILE64_SOURCE -D_FILE_OFFSET_BITS=64 -c -o hash.o hash.c\n",
      "gcc -O3 -std=c11 -pedantic -Wall -D_LARGEFILE64_SOURCE -D_FILE_OFFSET_BITS=64 -c -o color.o color.c\n",
      "gcc -O3 -std=c11 -pedantic -Wall -D_LARGEFILE64_SOURCE -D_FILE_OFFSET_BITS=64 -c -o file.o file.c\n",
      "gcc -O3 -std=c11 -pedantic -Wall -D_LARGEFILE64_SOURCE -D_FILE_OFFSET_BITS=64 -c -o filter.o filter.c\n",
      "gcc -O3 -std=c11 -pedantic -Wall -D_LARGEFILE64_SOURCE -D_FILE_OFFSET_BITS=64 -c -o info.o info.c\n",
      "gcc -O3 -std=c11 -pedantic -Wall -D_LARGEFILE64_SOURCE -D_FILE_OFFSET_BITS=64 -c -o unix.o unix.c\n",
      "gcc -O3 -std=c11 -pedantic -Wall -D_LARGEFILE64_SOURCE -D_FILE_OFFSET_BITS=64 -c -o xml.o xml.c\n",
      "gcc -O3 -std=c11 -pedantic -Wall -D_LARGEFILE64_SOURCE -D_FILE_OFFSET_BITS=64 -c -o json.o json.c\n",
      "gcc -O3 -std=c11 -pedantic -Wall -D_LARGEFILE64_SOURCE -D_FILE_OFFSET_BITS=64 -c -o html.o html.c\n",
      "gcc -O3 -std=c11 -pedantic -Wall -D_LARGEFILE64_SOURCE -D_FILE_OFFSET_BITS=64 -c -o strverscmp.o strverscmp.c\n",
      "gcc  -o tree tree.o list.o hash.o color.o file.o filter.o info.o unix.o xml.o json.o html.o strverscmp.o\n",
      "jq-linux64          100%[===================>]   3.77M  --.-KB/s    in 0.02s   \n",
      "Cloning into 'cellatlas'...\n",
      "remote: Enumerating objects: 267, done.\u001b[K\n",
      "remote: Counting objects: 100% (12/12), done.\u001b[K\n",
      "remote: Compressing objects: 100% (8/8), done.\u001b[K\n",
      "remote: Total 267 (delta 4), reused 11 (delta 4), pack-reused 255\u001b[K\n",
      "Receiving objects: 100% (267/267), 625.51 MiB | 20.22 MiB/s, done.\n",
      "Resolving deltas: 100% (133/133), done.\n",
      "Updating files: 100% (61/61), done.\n",
      "Proceed (Y/n)? "
     ]
    }
   ],
   "source": [
    "# Install `tree` to view files\n",
    "!wget --quiet --show-progress ftp://mama.indstate.edu/linux/tree/tree-2.1.0.tgz\n",
    "!tar -xf tree-2.1.0.tgz && cd tree-2.1.0 && make -j16 && make install > /dev/null\n",
    "\n",
    "# Install `jq, a command-line tool for extracting key value pairs from JSON files \n",
    "!wget --quiet --show-progress https://github.com/stedolan/jq/releases/download/jq-1.6/jq-linux64\n",
    "!chmod +x jq-linux64 && mv jq-linux64 /usr/local/bin/jq\n",
    "\n",
    "# Clone the cellatlas repo and install the package\n",
    "!git clone https://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@github.com/cellatlas/cellatlas.git > /dev/null\n",
    "!cd cellatlas && pip install --quiet . > /dev/null\n",
    "\n",
    "# Install dependencies\n",
    "!yes | pip uninstall --quiet seqspec\n",
    "!pip install --quiet git+https://github.com/IGVF/seqspec.git > /dev/null\n",
    "!pip install --quiet gget kb-python > /dev/null"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "6RhzGyP9_cEu"
   },
   "source": [
    "**Note:** We move the relevant data to the working directory and `gunzip` the barcode onlist."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "id": "vbIt2r2A3hbP"
   },
   "outputs": [],
   "source": [
    "!mv cellatlas/examples/rna-splitseq/* .\n",
    "!gunzip *.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "!gunzip *.gz"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "1ievY-64-u43"
   },
   "source": [
    "## Preprocessing"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "OrUgMsqp-weB"
   },
   "source": [
    "### Example the spec\n",
    "\n",
    "We first use `seqspec print` to check that the read structure matches what we expect. This command prints out an ordered tree representation of the sequenced elements contained in the FASTQ files. Note that the names of the nodes in the `seqspec` must match the names of the FASTQ files."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "oxK70t3e718z",
    "outputId": "0ef1385e-612f-42aa-c02e-edd31212d328",
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                      ┌─atac_I2.fastq.gz ──'barcode:16'\n",
      "                   ┌─atac─────────────┼─atac_R1.fastq.gz ──'gDNA-1:50'\n",
      "                   │                  └─atac_R2.fastq.gz ──'gDNA-2:50'\n",
      "───────────────────┤\n",
      "                   │                                     ┌─'barcode:16'\n",
      "                   │                  ┌─rna_R1.fastq.gz──┤\n",
      "                   └─rna──────────────┤                  └─'umi:12'\n",
      "                                      └─rna_R2.fastq.gz─ ──'cDNA:90'\n"
     ]
    }
   ],
   "source": [
    "!seqspec print spec.yaml"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "OExrOQ4f-xoy"
   },
   "source": [
    "### Fetch the references\n",
    "This step is only necessary if modality that we are processing uses a transcriptome reference-based alignment."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "I63OT0t92weV",
    "outputId": "85546a3f-6594-4753-da6f-02a24141ea18",
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Thu Jun  1 00:35:33 2023 INFO Fetching reference information for mus_musculus from Ensembl release: 109.\n"
     ]
    }
   ],
   "source": [
    "!gget ref -o ref.json -w dna,gtf mus_musculus"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "id": "xpHlsfEA8x1m",
    "tags": []
   },
   "outputs": [],
   "source": [
    "FA=!echo $(jq -r '.mus_musculus.genome_dna.ftp' ref.json)\n",
    "FA=FA[0]\n",
    "GTF=!echo $(jq -r '.mus_musculus.annotation_gtf.ftp' ref.json)\n",
    "GTF=GTF[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "hAa6kDU5-zak"
   },
   "source": [
    "### Build the pipeline\n",
    "\n",
    "We now supply all of the relevant objects to `cellatlas build` to produce the appropriate commands to be run to build the pipeline. This includes a reference building step and a read counting and quantification step both of which are performed with `kallisto` and `bustools` as part of the `kb-python` package."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "a.gz                 49%[========>           ] 379.30M   426KB/s    eta 15m 45s"
     ]
    }
   ],
   "source": [
    "!wget --quiet --show-progress $FA\n",
    "!wget --quiet --show-progress $GTF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {
    "id": "53NxlmVO2Ob_",
    "tags": []
   },
   "outputs": [],
   "source": [
    "!cellatlas build -o out \\\n",
    "-m atac \\\n",
    "-s spec.yaml \\\n",
    "-fa Mus_musculus.GRCm39.dna.primary_assembly.fa.gz -g Mus_musculus.GRCm39.109.gtf.gz \\\n",
    "fastqs/atac_R1.fastq.gz fastqs/atac_R2.fastq.gz fastqs/atac_I2.fastq.gz"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "2CvtZOUDAjfN"
   },
   "source": [
    "**Note**: The commands generated by `cellatlas build` are stored in the `out/cellatlas_info.json` file. We can view this file with `jq`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "049xtISb3sh4",
    "outputId": "0aa9b7c4-79d4-4cff-c3f3-24dd7752f9d8",
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "minimap2 -d out/ref.mmi Mus_musculus.GRCm39.dna.primary_assembly.fa.gz\n",
      "zcat Mus_musculus.GRCm39.dna.primary_assembly.fa.gz | fold -w 80 > out/genome.fa\n",
      "minimap2 -o out/genome.sam -a -x sr -t 32 out/ref.mmi fastqs/atac_R1.fastq.gz fastqs/atac_R2.fastq.gz\n",
      "samtools view -@ 8 -o out/genome.u.bam -b out/genome.sam\n",
      "samtools sort -@ 8 -o out/genome.bam -n -m 8G out/genome.u.bam\n",
      "Genrich -t out/genome.bam -o out/genome.bed -f out/genome_peaks.log -v\n",
      "cat out/genome.bed | bedtools sort | bedtools merge > out/peaks.bed\n",
      "bedtools getfasta -fi out/genome.fa -bed out/peaks.bed -fo out/peaks.fa\n",
      "cat out/peaks.fa | awk '{if($1~/>/)print $1\"\t\"$1\"\t\"$1}' > out/t2g.txt\n",
      "sed -i 's/>//g' out/t2g.txt\n",
      "kallisto index -i out/index.idx out/peaks.fa\n",
      "kb count -i out/index.idx -g out/t2g.txt -x 2,8,24:-1,-1,-1:0,0,50,1,0,50 -w ATA-737K-arc-v1.txt -o out --h5ad -t 2 fastqs/atac_R1.fastq.gz fastqs/atac_R2.fastq.gz fastqs/atac_I2.fastq.gz\n",
      "mkdir -p out/counts_mult\n",
      "bustools count -o out/counts_mult/cells_x_genes -g out/t2g.txt -e out/matrix.ec -t out/transcripts.txt --genecounts --cm out/output.unfiltered.bus\n"
     ]
    }
   ],
   "source": [
    "!jq  -r '.commands[] | values[] | join(\"\\n\")' out/cellatlas_info.json "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "FNidpYNN-10f"
   },
   "source": [
    "### Run the pipeline\n",
    "\n",
    "To run the pipeline we simply extract the commands from `out/cellatlas_info.json` and pass them to `bash`. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "7PDLF2Ps6wfz",
    "outputId": "d1fb3e02-3b51-4458-b248-89f1ce052ef7"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[M::mm_idx_gen::60.821*1.94] collected minimizers\n",
      "[M::mm_idx_gen::74.599*2.12] sorted minimizers\n",
      "[M::main::83.980*1.99] loaded/built the index for 61 target sequence(s)\n",
      "[M::mm_idx_stat] kmer size: 15; skip: 10; is_hpc: 0; #seq: 61\n",
      "[M::mm_idx_stat::84.846*1.98] distinct minimizers: 97791701 (39.85% are singletons); average occurrences: 5.142; average spacing: 5.426; total length: 2728222451\n",
      "[M::main] Version: 2.24-r1122\n",
      "[M::main] CMD: minimap2 -d out/ref.mmi Mus_musculus.GRCm39.dna.primary_assembly.fa.gz\n",
      "[M::main] Real time: 85.017 sec; CPU: 168.554 sec; Peak RSS: 10.621 GB\n",
      "[WARNING]\u001b[1;31m Indexing parameters (-k, -w or -H) overridden by parameters used in the prebuilt index.\u001b[0m\n",
      "[M::main::7.568*1.00] loaded/built the index for 61 target sequence(s)\n",
      "[M::mm_mapopt_update::7.568*1.00] mid_occ = 1000\n",
      "[M::mm_idx_stat] kmer size: 15; skip: 10; is_hpc: 0; #seq: 61\n",
      "[M::mm_idx_stat::8.646*1.00] distinct minimizers: 97791701 (39.85% are singletons); average occurrences: 5.142; average spacing: 5.426; total length: 2728222451\n",
      "[M::worker_pipeline::11.961*5.85] mapped 1010102 sequences\n",
      "[M::worker_pipeline::13.667*7.67] mapped 989898 sequences\n",
      "[M::main] Version: 2.24-r1122\n",
      "[M::main] CMD: minimap2 -o out/genome.sam -a -x sr -t 32 out/ref.mmi fastqs/atac_R1.fastq.gz fastqs/atac_R2.fastq.gz\n",
      "[M::main] Real time: 13.927 sec; CPU: 105.067 sec; Peak RSS: 7.352 GB\n",
      "[bam_sort_core] merging from 0 files and 8 in-memory blocks...\n",
      "Processing experimental file #0: out/genome.bam\n",
      "  BAM records analyzed:     2001003\n",
      "    Unmapped:               1926073\n",
      "    Supp./dups/lowQual:        1003\n",
      "    Paired alignments:        36842\n",
      "    Unpaired alignments:      37085\n",
      "  Fragments analyzed:         18421\n",
      "    Full fragments:           18421\n",
      "      (avg. length: 116.5bp)\n",
      "- control file #0 not provided -\n",
      "  Background pileup value: 0.000786\n",
      "Peak-calling parameters:\n",
      "  Genome length: 2728222451bp\n",
      "  Significance threshold: -log(p) > 2.000\n",
      "  Min. AUC: 200.000\n",
      "  Max. gap between sites: 100bp\n",
      "Peaks identified: 8234 (1177989bp)\n",
      "index file out/genome.fa.fai not found, generating...\n",
      "\n",
      "[build] loading fasta file out/peaks.fa\n",
      "[build] k-mer length: 31\n",
      "[build] warning: clipped off poly-A tail (longer than 10)\n",
      "        from 4 target sequences\n",
      "[build] counting k-mers ... done.\n",
      "[build] building target de Bruijn graph ...  done \n",
      "[build] creating equivalence classes ...  done\n",
      "[build] target de Bruijn graph has 11788 contigs and contains 888261 k-mers \n",
      "\n",
      "[2023-06-01 02:54:53,111]    INFO [count] Using index out/index.idx to generate BUS file to out from\n",
      "[2023-06-01 02:54:53,112]    INFO [count]         fastqs/atac_R1.fastq.gz\n",
      "[2023-06-01 02:54:53,112]    INFO [count]         fastqs/atac_R2.fastq.gz\n",
      "[2023-06-01 02:54:53,112]    INFO [count]         fastqs/atac_I2.fastq.gz\n",
      "[2023-06-01 02:54:59,203]    INFO [count] Sorting BUS file out/output.bus to out/tmp/output.s.bus\n"
     ]
    }
   ],
   "source": [
    "!jq  -r '.commands[] | values[] | join(\"\\n\")' out/cellatlas_info.json | bash"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "z5PcMU52A1DE"
   },
   "source": [
    "### Inspect outputs\n",
    "\n",
    "We inspect the `out/run_info.json` and `out/kb_info.json` as a simple QC on the pipeline."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "zCq2zEJD7cYf",
    "outputId": "04837748-5f78-4cb9-896a-b1d4e08b0fda"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{\n",
      "\t\"n_targets\": 0,\n",
      "\t\"n_bootstraps\": 0,\n",
      "\t\"n_processed\": 1000000,\n",
      "\t\"n_pseudoaligned\": 22885,\n",
      "\t\"n_unique\": 20194,\n",
      "\t\"p_pseudoaligned\": 2.3,\n",
      "\t\"p_unique\": 2.0,\n",
      "\t\"kallisto_version\": \"0.48.0\",\n",
      "\t\"index_version\": 0,\n",
      "\t\"start_time\": \"Thu Jun  1 02:42:59 2023\",\n",
      "\t\"call\": \"/home/sina/bin/miniconda3/lib/python3.7/site-packages/kb_python/bins/linux/kallisto/kallisto bus -i out/index.idx -o out -x 2,0,16:-1,-1,-1:0,0,50,1,0,50 -t 2 fastqs/atac_R1.fastq.gz fastqs/atac_R2.fastq.gz fastqs/atac_I2.fastq.gz\"\n",
      "}\n"
     ]
    }
   ],
   "source": [
    "!cat out/run_info.json"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "AS8SyOhmH7LB",
    "outputId": "5295e84d-1f61-49d6-d5df-da20f279dc5d"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "cat: out/kb_info.json: No such file or directory\n"
     ]
    }
   ],
   "source": [
    "!cat out/kb_info.json"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "H5mVaYkTIT4x",
    "outputId": "25ed138a-3dd8-4bc2-bc23-4315fec2cd8a"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[01;34mout\u001b[00m\n",
      "├── cellatlas_info.json\n",
      "├── \u001b[01;34mcounts_mult\u001b[00m\n",
      "├── genome.bam\n",
      "├── genome.bed\n",
      "├── genome.fa\n",
      "├── genome.fa.fai\n",
      "├── genome_peaks.log\n",
      "├── genome.sam\n",
      "├── genome.u.bam\n",
      "├── index.idx\n",
      "├── index.saved\n",
      "├── inspect.json\n",
      "├── matrix.ec\n",
      "├── output.bus\n",
      "├── peaks.bed\n",
      "├── peaks.fa\n",
      "├── ref.mmi\n",
      "├── run_info.json\n",
      "├── t2g.txt\n",
      "└── transcripts.txt\n",
      "\n",
      "1 directory, 19 files\n"
     ]
    }
   ],
   "source": [
    "!tree out"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "colab": {
   "authorship_tag": "ABX9TyOpJNaSEgTr3/2KscpJzhQV",
   "include_colab_link": true,
   "provenance": []
  },
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
