{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": [],
      "authorship_tag": "ABX9TyMaz/WkRboaLhc5HnRtyKAr",
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/cellatlas/cellatlas/blob/main/examples/rna-10xv3/preprocess.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!wget --quiet --show-progress https://github.com/stedolan/jq/releases/download/jq-1.6/jq-linux64\n",
        "!chmod +x jq-linux64 && mv jq-linux64 /usr/local/bin/jq"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "DCtbJUnW6Qta",
        "outputId": "e65f3dbf-4eb4-4afd-cb2f-7b9f1de46a51"
      },
      "execution_count": 1,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\rjq-linux64            0%[                    ]       0  --.-KB/s               \rjq-linux64          100%[===================>]   3.77M  --.-KB/s    in 0.05s   \n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 2,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "OEii54Z91IKD",
        "outputId": "558e2005-2e96-4b6d-b76e-6a731aa41ba7"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Cloning into 'cellatlas'...\n",
            "remote: Enumerating objects: 153, done.\u001b[K\n",
            "remote: Counting objects: 100% (28/28), done.\u001b[K\n",
            "remote: Compressing objects: 100% (22/22), done.\u001b[K\n",
            "remote: Total 153 (delta 9), reused 21 (delta 6), pack-reused 125\u001b[K\n",
            "Receiving objects: 100% (153/153), 527.01 MiB | 25.51 MiB/s, done.\n",
            "Resolving deltas: 100% (48/48), done.\n",
            "Updating files: 100% (51/51), done.\n"
          ]
        }
      ],
      "source": [
        "!git clone https://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@github.com/cellatlas/cellatlas.git > /dev/null"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!cd cellatlas && pip install --quiet . > /dev/null\n",
        "!pip install --quiet git+https://github.com/IGVF/seqspec.git > /dev/null\n",
        "!pip install --quiet gget kb-python > /dev/null"
      ],
      "metadata": {
        "id": "2VTXQLMD2Y6m"
      },
      "execution_count": 3,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "!mv cellatlas/examples/rna-10xv3/* .\n",
        "!gunzip 3M-february-2018.txt.gz"
      ],
      "metadata": {
        "id": "vbIt2r2A3hbP"
      },
      "execution_count": 4,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Preprocessing"
      ],
      "metadata": {
        "id": "1ievY-64-u43"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "### Example the spec"
      ],
      "metadata": {
        "id": "OrUgMsqp-weB"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!seqspec print spec.yaml"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "oxK70t3e718z",
        "outputId": "4ccaa615-a004-4fa4-9097-fa3f5988de24"
      },
      "execution_count": 5,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "                                             ┌─'barcode:16'\n",
            "                              ┌─R1.fastq.gz──┤\n",
            "────────────── ──rna──────────┤              └─'umi:12'\n",
            "                              └─R2.fastq.gz─ ──'cDNA:90'\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "### Fetch the references"
      ],
      "metadata": {
        "id": "OExrOQ4f-xoy"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!gget ref -o ref.json -w dna,gtf homo_sapiens"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "I63OT0t92weV",
        "outputId": "30a38ad5-3228-4446-81f2-f940e853a307"
      },
      "execution_count": 6,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Tue May  9 00:52:50 2023 INFO Fetching reference information for homo_sapiens from Ensembl release: 109.\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "FA=!echo $(jq -r '.homo_sapiens.genome_dna.ftp' ref.json)\n",
        "FA=FA[0]\n",
        "GTF=!echo $(jq -r '.homo_sapiens.annotation_gtf.ftp' ref.json)\n",
        "GTF=GTF[0]"
      ],
      "metadata": {
        "id": "xpHlsfEA8x1m"
      },
      "execution_count": 7,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "### Build the pipeline"
      ],
      "metadata": {
        "id": "hAa6kDU5-zak"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!cellatlas build -o out \\\n",
        "-m rna \\\n",
        "-s spec.yaml \\\n",
        "-fa $FA -g $GTF \\\n",
        "fastqs/R1.fastq.gz fastqs/R2.fastq.gz"
      ],
      "metadata": {
        "id": "53NxlmVO2Ob_"
      },
      "execution_count": 8,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "!jq  -r '.commands[] | values[] | join(\"\\n\")' out/cellatlas_info.json "
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "049xtISb3sh4",
        "outputId": "6472e4fe-c029-43b0-fce6-49053f16b818"
      },
      "execution_count": 9,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "kb ref -i out/index.idx -g out/t2g.txt -f1 out/transcriptome.fa http://ftp.ensembl.org/pub/release-109/fasta/homo_sapiens/dna/Homo_sapiens.GRCh38.dna.primary_assembly.fa.gz http://ftp.ensembl.org/pub/release-109/gtf/homo_sapiens/Homo_sapiens.GRCh38.109.gtf.gz\n",
            "kb count -i out/index.idx -g out/t2g.txt -x 0,0,16:0,16,28:1,0,90 -w 3M-february-2018.txt -o out/output --h5ad -t 2 fastqs/R1.fastq.gz fastqs/R2.fastq.gz\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "### Run the pipeline"
      ],
      "metadata": {
        "id": "FNidpYNN-10f"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!jq  -r '.commands[] | values[] | join(\"\\n\")' out/cellatlas_info.json | bash"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "7PDLF2Ps6wfz",
        "outputId": "b244ad99-7482-4fc1-aaf3-066e11f9047e"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "[2023-05-09 00:52:59,078]    INFO [ref] Preparing http://ftp.ensembl.org/pub/release-109/fasta/homo_sapiens/dna/Homo_sapiens.GRCh38.dna.primary_assembly.fa.gz, http://ftp.ensembl.org/pub/release-109/gtf/homo_sapiens/Homo_sapiens.GRCh38.109.gtf.gz\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "zCq2zEJD7cYf"
      },
      "execution_count": null,
      "outputs": []
    }
  ]
}