{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": [],
      "authorship_tag": "ABX9TyMaz/WkRboaLhc5HnRtyKAr",
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/cellatlas/cellatlas/blob/main/examples/rna-10xv3/preprocessing.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!wget --quiet --show-progress https://github.com/stedolan/jq/releases/download/jq-1.6/jq-linux64\n",
        "!chmod +x jq-linux64 && mv jq-linux64 /usr/local/bin/jq"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "DCtbJUnW6Qta",
        "outputId": "fe0a6a61-6a64-403a-d9c4-7b70c97ca0fb"
      },
      "execution_count": 1,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\rjq-linux64            0%[                    ]       0  --.-KB/s               \rjq-linux64          100%[===================>]   3.77M  --.-KB/s    in 0.07s   \n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 2,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "OEii54Z91IKD",
        "outputId": "7585025b-fc5f-4378-cd72-932af93fef78"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Cloning into 'cellatlas'...\n",
            "remote: Enumerating objects: 158, done.\u001b[K\n",
            "remote: Counting objects: 100% (33/33), done.\u001b[K\n",
            "remote: Compressing objects: 100% (27/27), done.\u001b[K\n",
            "remote: Total 158 (delta 13), reused 21 (delta 6), pack-reused 125\u001b[K\n",
            "Receiving objects: 100% (158/158), 527.01 MiB | 40.89 MiB/s, done.\n",
            "Resolving deltas: 100% (52/52), done.\n",
            "Updating files: 100% (51/51), done.\n"
          ]
        }
      ],
      "source": [
        "!git clone https://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@github.com/cellatlas/cellatlas.git > /dev/null"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!cd cellatlas && pip install --quiet . > /dev/null\n",
        "!pip install --quiet git+https://github.com/IGVF/seqspec.git > /dev/null\n",
        "!pip install --quiet gget kb-python > /dev/null"
      ],
      "metadata": {
        "id": "2VTXQLMD2Y6m"
      },
      "execution_count": 3,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "!mv cellatlas/examples/rna-10xv3/* .\n",
        "!gunzip 3M-february-2018.txt.gz"
      ],
      "metadata": {
        "id": "vbIt2r2A3hbP"
      },
      "execution_count": 4,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Preprocessing"
      ],
      "metadata": {
        "id": "1ievY-64-u43"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "### Example the spec"
      ],
      "metadata": {
        "id": "OrUgMsqp-weB"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!seqspec print spec.yaml"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "oxK70t3e718z",
        "outputId": "5bd60724-6304-4dae-dfc1-21e9a31c8e7d"
      },
      "execution_count": 5,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "                                             ┌─'barcode:16'\n",
            "                              ┌─R1.fastq.gz──┤\n",
            "────────────── ──rna──────────┤              └─'umi:12'\n",
            "                              └─R2.fastq.gz─ ──'cDNA:90'\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "### Fetch the references"
      ],
      "metadata": {
        "id": "OExrOQ4f-xoy"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!gget ref -o ref.json -w dna,gtf homo_sapiens"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "I63OT0t92weV",
        "outputId": "b1532597-7664-4494-bafc-2ab9dc5c199e"
      },
      "execution_count": 6,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Tue May  9 04:44:14 2023 INFO Fetching reference information for homo_sapiens from Ensembl release: 109.\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "FA=!echo $(jq -r '.homo_sapiens.genome_dna.ftp' ref.json)\n",
        "FA=FA[0]\n",
        "GTF=!echo $(jq -r '.homo_sapiens.annotation_gtf.ftp' ref.json)\n",
        "GTF=GTF[0]"
      ],
      "metadata": {
        "id": "xpHlsfEA8x1m"
      },
      "execution_count": 7,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "### Build the pipeline"
      ],
      "metadata": {
        "id": "hAa6kDU5-zak"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!cellatlas build -o out \\\n",
        "-m rna \\\n",
        "-s spec.yaml \\\n",
        "-fa $FA -g $GTF \\\n",
        "fastqs/R1.fastq.gz fastqs/R2.fastq.gz"
      ],
      "metadata": {
        "id": "53NxlmVO2Ob_"
      },
      "execution_count": 8,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "!jq  -r '.commands[] | values[] | join(\"\\n\")' out/cellatlas_info.json "
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "049xtISb3sh4",
        "outputId": "e71ea412-16a1-46c4-8660-d46f244f0c26"
      },
      "execution_count": 9,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "kb ref -i out/index.idx -g out/t2g.txt -f1 out/transcriptome.fa http://ftp.ensembl.org/pub/release-109/fasta/homo_sapiens/dna/Homo_sapiens.GRCh38.dna.primary_assembly.fa.gz http://ftp.ensembl.org/pub/release-109/gtf/homo_sapiens/Homo_sapiens.GRCh38.109.gtf.gz\n",
            "kb count -i out/index.idx -g out/t2g.txt -x 0,0,16:0,16,28:1,0,90 -w 3M-february-2018.txt -o out/output --h5ad -t 2 fastqs/R1.fastq.gz fastqs/R2.fastq.gz\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "### Run the pipeline"
      ],
      "metadata": {
        "id": "FNidpYNN-10f"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!jq  -r '.commands[] | values[] | join(\"\\n\")' out/cellatlas_info.json | bash"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "7PDLF2Ps6wfz",
        "outputId": "b8abb08c-ae92-4c77-8d8c-cfa52f7e619d"
      },
      "execution_count": 10,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "[2023-05-09 04:44:24,318]    INFO [ref] Preparing http://ftp.ensembl.org/pub/release-109/fasta/homo_sapiens/dna/Homo_sapiens.GRCh38.dna.primary_assembly.fa.gz, http://ftp.ensembl.org/pub/release-109/gtf/homo_sapiens/Homo_sapiens.GRCh38.109.gtf.gz\n",
            "[2023-05-09 04:46:15,474]    INFO [ref] Splitting genome http://ftp.ensembl.org/pub/release-109/fasta/homo_sapiens/dna/Homo_sapiens.GRCh38.dna.primary_assembly.fa.gz into cDNA at /content/tmp/tmpybpilki4\n",
            "[2023-05-09 05:07:32,542]    INFO [ref] Concatenating 1 cDNAs to out/transcriptome.fa\n",
            "[2023-05-09 05:07:34,588]    INFO [ref] Creating transcript-to-gene mapping at out/t2g.txt\n",
            "[2023-05-09 05:07:38,683]    INFO [ref] Indexing out/transcriptome.fa to out/index.idx\n",
            "[2023-05-09 05:20:35,534]    INFO [count] Using index out/index.idx to generate BUS file to out/output from\n",
            "[2023-05-09 05:20:35,534]    INFO [count]         fastqs/R1.fastq.gz\n",
            "[2023-05-09 05:20:35,534]    INFO [count]         fastqs/R2.fastq.gz\n",
            "[2023-05-09 05:21:27,890]    INFO [count] Sorting BUS file out/output/output.bus to out/output/tmp/output.s.bus\n",
            "[2023-05-09 05:21:34,306]    INFO [count] Inspecting BUS file out/output/tmp/output.s.bus\n",
            "[2023-05-09 05:21:40,424]    INFO [count] Correcting BUS records in out/output/tmp/output.s.bus to out/output/tmp/output.s.c.bus with whitelist 3M-february-2018.txt\n",
            "[2023-05-09 05:21:56,562]    INFO [count] Sorting BUS file out/output/tmp/output.s.c.bus to out/output/output.unfiltered.bus\n",
            "[2023-05-09 05:22:02,079]    INFO [count] Generating count matrix out/output/counts_unfiltered/cells_x_genes from BUS file out/output/output.unfiltered.bus\n",
            "[2023-05-09 05:22:08,186]    INFO [count] Reading matrix out/output/counts_unfiltered/cells_x_genes.mtx\n",
            "[2023-05-09 05:22:09,639] WARNING [count] 20930 gene IDs do not have corresponding gene names. These genes will use their gene IDs instead.\n",
            "[2023-05-09 05:22:09,669]    INFO [count] Writing matrix to h5ad out/output/counts_unfiltered/adata.h5ad\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "zCq2zEJD7cYf"
      },
      "execution_count": 10,
      "outputs": []
    }
  ]
}